File: app/src/main/java/com/rdev/tryp/blocks/screens/notifications/NotificationsAdapter.java
package com.rdev.tryp.blocks.screens.notifications;

import android.content.Context;
import android.graphics.Typeface;
import android.text.Spannable;
import android.text.SpannableString;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.rdev.tryp.R;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import io.github.inflationx.calligraphy3.CalligraphyTypefaceSpan;
import io.github.inflationx.calligraphy3.TypefaceUtils;

/**
 * Created by Alexey Matrosov on 03.03.2019.
 */
public class NotificationsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final int TYPE_TEXT = 0;
    private static final int TYPE_PROMO = 1;

    private List<NotificationItem> items;
    private Context context;

    public NotificationsAdapter(Context context, List<NotificationItem> items) {
        this.context = context;
        this.items = items;
    }

    @Override
    public int getItemViewType(int position) {
        NotificationItem item = items.get(position);
        switch (item.getType()) {
            case Promo: return TYPE_PROMO;
            case Normal: return  TYPE_TEXT;
            default: throw new IllegalStateException("Unknown item type");
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case TYPE_TEXT: {
                View view = LayoutInflater.from(context).inflate(R.layout.item_notificatio_text, parent, false);
                return new TextViewHolder(view);
            }

            case TYPE_PROMO: {
                View view = LayoutInflater.from(context).inflate(R.layout.item_notification_promo, parent, false);
                return new PromoViewHolder(view);
            }
            default: throw new IllegalStateException("Unknown item type");
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        switch (holder.getItemViewType()) {
            case TYPE_TEXT: {
                TextViewHolder textHolder = (TextViewHolder) holder;
                NotificationItem item = items.get(position);

                textHolder.name.setText(item.getName());
                textHolder.message.setText(item.getMessage());
                textHolder.userIcon.setImageResource(item.getUserIcon() == 0 ? R.drawable.invite_default_icon : item.getUserIcon());
                textHolder.time.setText(item.getTime());

                break;
            }

            case TYPE_PROMO: {
                PromoViewHolder promoHolder = (PromoViewHolder) holder;
                NotificationItem item = items.get(position);

                promoHolder.promoImage.setImageResource(item.getPhomoImage());
                promoHolder.title.setText(item.getTitle());
                promoHolder.time.setText(item.getTime());
                changeCodePart(promoHolder.description, item.getDescription(), item.getCode());

                break;
            }
            default: throw new IllegalStateException("Unknown item type");
        }
    }

    @Override
    public int getItemCount() {
        return items == null ? 0 : items.size();
    }

    private void changeCodePart(TextView description, String text, String code) {
        int coloredPart1Index = text.indexOf(code);

        Spannable spannable = new SpannableString(text);
        if (coloredPart1Index != -1 && context != null) {
            Typeface typeface = TypefaceUtils.load(context.getAssets(), "fonts/OpenSans-Bold.ttf");

            spannable.setSpan(new CalligraphyTypefaceSpan(typeface),
                    coloredPart1Index,
                    coloredPart1Index + code.length(),
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        }

        description.setText(spannable, TextView.BufferType.SPANNABLE);
    }

    class TextViewHolder extends RecyclerView.ViewHolder {
        ImageView userIcon;
        TextView name;
        TextView message;
        TextView time;

        public TextViewHolder(@NonNull View itemView) {
            super(itemView);

            userIcon = itemView.findViewById(R.id.notification_icon_container);
            name = itemView.findViewById(R.id.item_notification_title);
            message = itemView.findViewById(R.id.item_notification_message);
            time = itemView.findViewById(R.id.notification_time);
        }
    }

    class PromoViewHolder extends RecyclerView.ViewHolder {
        ImageView promoImage;
        TextView title;
        TextView description;
        TextView time;

        public PromoViewHolder(@NonNull View itemView) {
            super(itemView);

            promoImage = itemView.findViewById(R.id.notification_promo);
            title = itemView.findViewById(R.id.notification_promo_title);
            description = itemView.findViewById(R.id.notification_promo_description);
            time = itemView.findViewById(R.id.notification_time);
        }
    }
}

File: app/src/main/java/com/rdev/tryp/blocks/screens/notifications/NotificationsAdapter.kt
package com.rdev.tryp.blocks.screens.notifications

import android.content.Context
import android.text.Spannable
import android.text.SpannableString
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.rdev.tryp.R
import androidx.recyclerview.widget.RecyclerView
import io.github.inflationx.calligraphy3.CalligraphyTypefaceSpan
import io.github.inflationx.calligraphy3.TypefaceUtils

/**
 * Created by Alexey Matrosov on 03.03.2019.
 */


class NotificationsAdapter(private val context: Context?, private val items: List<NotificationItem>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        const val TYPE_TEXT = 0
        const val TYPE_PROMO = 1
    }

    override fun getItemViewType(position: Int) = when (items[position].type) {
        NotificationItem.Type.Promo -> TYPE_PROMO
        NotificationItem.Type.Normal -> TYPE_TEXT
        else -> throw IllegalStateException("Unknown item type")
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = when (viewType) {
        TYPE_TEXT -> {
            val view = LayoutInflater.from(context).inflate(R.layout.item_notificatio_text, parent, false)
            TextViewHolder(view)
        }

        TYPE_PROMO -> {
            val view = LayoutInflater.from(context).inflate(R.layout.item_notification_promo, parent, false)
            PromoViewHolder(view)
        }
        else -> throw IllegalStateException("Unknown item type")
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            TYPE_TEXT -> {
                val textHolder = holder as TextViewHolder
                val item = items[position]

                textHolder.name.text = item.name
                textHolder.message.text = item.message
                textHolder.userIcon.setImageResource(if (item.userIcon == 0) R.drawable.invite_default_icon else item.userIcon)
                textHolder.time.text = item.time
            }

            TYPE_PROMO -> {
                val promoHolder = holder as PromoViewHolder
                val item = items[position]

                promoHolder.promoImage.setImageResource(item.phomoImage)
                promoHolder.title.text = item.title
                promoHolder.time.text = item.time
                item.description?.let { description ->
                    item.code?.let { code ->
                        changeCodePart(promoHolder.description, description, code)
                    }
                }
            }
            else -> throw IllegalStateException("Unknown item type")
        }
    }

    override fun getItemCount() = items.size

    private fun changeCodePart(description: TextView, text: String, code: String) {
        val coloredPart1Index = text.indexOf(code)

        val spannable = SpannableString(text)
        if (coloredPart1Index != -1 && context != null) {
            val typeface = TypefaceUtils.load(context.assets, "fonts/OpenSans-Bold.ttf")

            spannable.setSpan(CalligraphyTypefaceSpan(typeface),
                    coloredPart1Index,
                    coloredPart1Index + code.length,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        }

        description.setText(spannable, TextView.BufferType.SPANNABLE)
    }

    inner class TextViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var userIcon: ImageView = itemView.findViewById(R.id.notification_icon_container)
        var name: TextView = itemView.findViewById(R.id.item_notification_title)
        var message: TextView = itemView.findViewById(R.id.item_notification_message)
        var time: TextView = itemView.findViewById(R.id.notification_time)
    }

    inner class PromoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var promoImage: ImageView = itemView.findViewById(R.id.notification_promo)
        var title: TextView = itemView.findViewById(R.id.notification_promo_title)
        var description: TextView = itemView.findViewById(R.id.notification_promo_description)
        var time: TextView = itemView.findViewById(R.id.notification_time)
    }

}